On-demand log-spacemap flush; `zpool condense` command

robn opened the pull request:
_[Sponsors: TrueNAS, Klara, Inc., Wasabi Technology, Inc.]_

### Motivation and Context

Normally, log spacemaps are flushed out to the metaslabs when the pool is exported. For large logs, this can lead to export taking an inordinate amount of time.

This PR adds a an on-demand variant of the log spacemap flush, and a `zpool condense`  command to trigger it. With it, an operator can request that log spacemaps be flushed ahead of time, so that there is relatively little work to be done at export time.

### Description

There's two halves to this.

First, we add a "mode" to the existing "flushall" behaviour in `spa_flush_metaslabs()`. The traditional behaviour is now "export mode", and flushes all logs. Some new functions are added to start and stop the flush with a given mode. Then we add a "request" mode, for use by the operator. This follows the same logic of walking the logs and flushing them out, but skips any that were modified after the flush request was made.

The second part is the addition of the `zpool condense` command, and support library and ioctl additions. This takes a `-t <target>` parameter, which is the "thing" to condense, flush, garbage-collect or otherwise accelerate background processing for. It's designed so it could be wired up to any similar background process in the future. In particular, I had the dedup log in mind while putting it together.

All the trimmings you'd expect are there. Condense operations can be cancelled, restoring the flushing behaviour to its original pace or schedule. They can be waited on, via `condense -w` or `wait -t condense`. The latter combines all condense targets into one signal value, theoretically allowing multiple things to be condensed at the same time, and wait until they're all finished. Without a second or maybe even third target it's unclear to me if this it what the user will expect, but I don't think this is far off and it can be changes when the next thing gets hooked up.

I've included a kstat exposing the pool "unflushed" counters. We used this in our initial investigations. I thought it was going to be useful for the ZTS test, but it ended up being too difficult to control reliably. So nothing here uses it directly, but it doesn't hurt anything to have it there and may help someone, so why not.

### How Has This Been Tested?

`ztest` gets support, and has had many tens of runs without issue. ZTS test has been added, and the entire suite run to successful completion.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

robn committed:
spa_stats: kstats for unflushed log spacemaps

Sponsored-by: TrueNAS
Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn committed:
log_spacemap: extend pool flushall to have "request" and "export" modes

Normally, log spacemaps are flushed out to the metaslabs when the pool
is exported. For large logs, this can lead to export taking an
inordinate amount of time.

This commit adds a "mode" parameter for the log spacemap "flushall"
operation, and functions for starting and stopping it in a particular
mode. The existing behaviour of flushing everything is now the "export"
mode.

Then, we add a new "request" mode, that can be triggered externally.
This activates the same flushall code, with a few differences:

- we only consider flushing metaslabs that were dirtied on the txg
  before the flushall operation was started

- we close and issue the txg immediately when the flushall is active,
  rather than wait for zfs_txg_timeout each time (similar to how scrub
  also forces the txg forward)

- each txg, we flush a fixed number of metaslabs, regardless of the rate
  of change. The amount is set by a new tunable new tunable
  zfs_min_metaslabs_to_condense_pct, which defaults to 5% of the number
  of unflushed metaslabs at the time the condense operation was started.

The minimum dirty metaslabs pct% together with forcing the txg forward
should mean that with the defaults, condensing is attempted at 5% per
second, and so on an idle pool should take no more than 20s to flush.

Sponsored-by: TrueNAS
Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn committed:
zpool: add condense verb

The idea is to have a single command that could signal to any background
cleanup task that it should do its work faster, or care less about not
getting in the way of user IO, or whatever.

This adds the the `zpool condense` command, the `ZFS_IOC_POOL_CONDENSE`
ioctl and counters so userspace can get progress. Because the type could
be anything, there's no particular unit, just a total number of items to
condense and count of how many done.

Included is a "debug" condense type. In debug builds, issuing condense
with this type will start a background process that will simply bump the
condense counters every second for ten seconds. This is intended for use
by the test suite and for debugging the condense infrastructure itself,
and will be compiled out production builds.

Sponsored-by: TrueNAS
Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn committed:
ztest: periodically start/stop log spacemap flush

Sponsored-by: TrueNAS
Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn committed:
zts: add test for log spacemap flushall + zpool condense

Sponsored-by: TrueNAS
Sponsored-by: Klara, Inc.
Sponsored-by: Wasabi Technology, Inc.
Signed-off-by: Rob Norris <rob.norris@truenas.com>

amotin reviewed (commented):
May be I've missed something, but what will make transactions move and more metaslabs flushed if the pool is idle?

Also, if the pool is idle, will it flush only 5 transactions per transaction? I worry about the amount of new dirty metaslabs/transactions it may produce until it finally converge.

robn commented on cmd/zpool/zpool_main.c:
```diff
@@ -8688,6 +8704,122 @@ zpool_do_trim(int argc, char **argv)
+		}
+	}
+
+	if (cb.type == POOL_CONDENSE_TYPES) {
+		(void) fprintf(stderr, gettext("missing condense target\n"));
```
Not sure! I figure we could add it when the second one comes along. Comma-separated might be nice, like wait targets etc.

robn commented on cmd/zpool/zpool_main.c:
```diff
@@ -8688,6 +8704,122 @@ zpool_do_trim(int argc, char **argv)
+	argc -= optind;
+	argv += optind;
+
+	if (argc < 1) {
+		(void) fprintf(stderr, gettext("missing pool name argument\n"));
```
Yeah, I didn't want an "all pools" mode here, because I don't know what might be there in the future, and it might mean different things for different pools. Same way `zpool scrub` requires a pool arg.

robn commented on module/zfs/spa_log_spacemap.c:
```diff
@@ -746,10 +755,87 @@ spa_log_exceeds_memlimit(spa_t *spa)
-boolean_t
-spa_flush_all_logs_requested(spa_t *spa)
+void
+spa_log_flushall_start(spa_t *spa, spa_log_flushall_mode_t mode, uint64_t txg)
```
Ahh, maybe I wasn't clear. These are just the controls for log spacemap flushall. The next thing would be something separate, just triggered from `zfs_ioc_condense()`.

amotin replied:
I am not saying that this function will handle other logs. Just that it is not very specific for externally visible name. Even more for `spa_log_flushall_mode` actually. But looking now there are other names around it that are also not metaslab-specific, so I don't insist.

robn replied:
Ahh, I see what you mean. Yeah, just following the surrounding style... :/

robn commented:
> May be I've missed something, but what will make transactions move and more metaslabs flushed if the pool is idle?
> 
> Also, if the pool is idle, will it flush only 5 transactions per transaction? I worry about the amount of new dirty metaslabs/transactions it may produce until it finally converge.

@amotin mm, you may be right. It didn't come up in testing, but we hadn't gone out of our way to stop pool activity. (Also I wrote this last year, so probably didn't know about this at the time!)

I'll study it and post an update soon, thanks!

robn commented:
Right, I think I've swapped back in everything I need.

So yes, you're right - when the pool is idle, nothing is pushing things along (same for dedup log, incidentally). The 5s timeout will see some flushed out, but that's all. This is existing behaviour, so I'm ok with that, I think.

So it seems to me that there's two questions.

If the operator has requested spacemap log flush, should we push the sync along a bit? I think it's reasonable to say yes, in theory. Similar idea to the `dsl_scan_active()` call in `txg_sync_thread()`.

Then, what should the amount be. I forget why we chose minimum 5; probably it was just a number that was easy to see vs 1. Of course, we should use more the quieter things are. Is that just a much higher minimum? Maybe a % of the total, or of the amount dirty? Or is it more like, some larger minimum + inverse of the change rate, so we always get a big amount, and more if there's space to do it.

I'll ask around. Let me know if you have any thoughts.

amotin commented:
@robn I think once user requested condense, we should do it as fast as possible, since user likely waits for it to reboot, export, whatever.  This operation makes no sense to do just routinely.  The only limitation is not hurt other workload too much.

The amount of sync is a good question.  I'd guess we don't want to extend transaction group for too long, neither consume to much memory on dirty data, etc.  But my last trip to spacemaps was some time ago, so I don't have specific recommendations.

robn commented on cmd/zpool/zpool_main.c:
```diff
@@ -8845,6 +8862,105 @@ zpool_do_trim(int argc, char **argv)
+zpool_do_condense(int argc, char **argv)
+{
+	struct option long_options[] = {
+		{"target",	required_argument,	NULL,	't'},
```
Yeah nice. Also added `-a` for "all pools"; there's precedent now that `scrub` etc has it.

robn committed:
ABI: bump for spacemap condense

Sponsored-by: TrueNAS
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn committed:
log_spacemap: add counter for unflushed metaslabs

Useful for understanding, but also a convenient place to grab the
current count without needing to walk over the metaslab list.

Sponsored-by: TrueNAS
Signed-off-by: Rob Norris <rob.norris@truenas.com>

amotin commented on cmd/zpool/zpool_main.c:
```diff
@@ -9929,6 +10075,55 @@ removal_status_nvlist(zpool_handle_t *zhp, status_cbdata_t *cb,
+		    ZFS_NICENUM_1024);
+		fnvlist_add_string(nv, "unit", condense_type_unit_str[type]);
+
+		fnvlist_add_nvlist(cnv, condense_type_nv_str[type], nv);
```
If we add a new type at some point, I suppose old binaries will crash here getting our of array.  Below in `print_condense_status()` you seem to handle it.  I wonder if kernel could provide more information, so that user-space would need less external knowledge.

robn replied:
Should be handled a few lines up, `uint_t n = MIN(POOL_CONDENSE_TYPES, ...)`. So it should be crash safe.

Thinking more, I wonder if instead of using an array, it should be another kv, eg `{"log_spacemap":{...}}` (basically exactly the JSON form here). Then userspace can only display the ones it knows about, and if we ever decided to remove one in the future, the kernel can just not fill it out, and userspace just won't do anything with it, vs trying to handle an empty element as is the case now.

Yeah, I definitely prefer that, I'll get right on it!

amotin commented on module/zfs/spa_stats.c:
```diff
@@ -1037,6 +1042,86 @@ spa_iostats_destroy(spa_t *spa)
+	SPA_LOG_SM_STATS_SET(unflushed_blocklimit,
+	    spa->spa_unflushed_stats.sus_blocklimit);
+	SPA_LOG_SM_STATS_SET(unflushed_nblocks,
+	    spa->spa_unflushed_stats.sus_nblocks);
```
Not a bug, but so much infrastructure just to have stats copied atomically...

robn replied:
I swear I have a branch that has all this stats stuff reduced to a few macros. I mention it every time it comes up, I vividly remember doing it, I just haven't found it yet. Maybe I dreamed it? I should just go do it one way or another.

(I was just thinking, "I distinctly remember writing it on the flight home from Toronto in 2024", but that trip was chaos due to weather in SFO. So maybe it _was_ a fever/exhaustion dream...)

amotin commented on module/zfs/spa_log_spacemap.c:
```diff
@@ -747,10 +756,116 @@ spa_log_exceeds_memlimit(spa_t *spa)
+		pcns->pcns_total = spa_log_sm_unflushed_metaslabs(spa);
+
+		/* Get it started immediately. */
+		txg_kick(spa->spa_dsl_pool, txg + 1);
```
I about not sure about use of the `txg` here.  From one side if represents a TXG up to which we want to flush, and that is fine.  But form another, here it is used to start the sync process, and if we don't have the transaction with this number help open (as in case when 0 was passed in), I worry here might be a race if the TXG is already being synced.  Am I wrong?

robn replied:
Best I can tell its just a signalling mechanism, telling `txg_sync()` to keep going until this txg is out. `txg_kick` is just setting `tx_sync_txg_waiting` and then `cv_broadcast()` under lock.  If its already being synced out, then I think `tx_sync_txg_waiting` just won't be changed. If there's a race possible (unsure, but there are a couple of unlocked reads) then I think at worst it'll just end up waiting the full txg_timeout and then getting started?

amotin replied:
I think in `txg_kick()` just should not be the same TXG as passed into `spa_log_flushall_start()`.  The last one means up to which TXG to sync, which I can speculate being in some past.  The first one though should always be a TXG that is going to be synced next.  Though in case of call from `spa_unload_log_sm_flush_all()` the semantic is the only right one of available.

I agree about "worst it'll just end up waiting the full txg_timeout", so it should not be fatal, only a few seconds delay for no reason, but it does not look right.

robn replied:
Fair enough. I've made it `spa_syncing_txg(spa) + 1`, which should at least ensure one round. The only other option that looks halfway plausible is `spa_open_txg(spa) + TXG_DEFER_SIZE`, which is (effectively) what `txg_wait_synced(0)` does. But I think I don't really like that, because we're just trying to push out logs, getting open context out too is kinda not the point?

amotin replied:
OK. `TXG_DEFER_SIZE` has nothing to do with this use case, or actually most cases that don't need final space accounting after some operation.

amotin commented on module/zfs/spa_log_spacemap.c:
```diff
@@ -804,17 +919,38 @@ spa_flush_metaslabs(spa_t *spa, dmu_tx_t *tx)
+		pool_condense_stat_t *pcns =
+		    &spa->spa_condense_stats[POOL_CONDENSE_LOG_SPACEMAP];
+		want_to_flush = MAX(pcns->pcns_total, pcns->pcns_total *
+		    MAX(MIN(zfs_min_metaslabs_to_condense_pct, 100), 1) / 100);
```
`MAX(pcns->pcns_total, pcns->pcns_total * pct / 100)` does not make sense to me, since it will always be equal to `pcns->pcns_total`.  Am I missing something?

Also I wonder if we should never flush less than `spa_estimate_metaslabs_to_flush(spa)`, or we may never converge and just make things worse.

robn replied:
Hurr, yeah, I think it's some dumb holdover from the previous revision. It could plausibly be `MIN` instead of `MAX`, but that's implied with the 100% cap anyway.

And yes, agreed on `spa_estimate_metaslabs_to_flush`, since that's kind of our "do nothing" state. We should never do worse than that!

amotin replied:
This is better, just not sure I like `spa_estimate_metaslabs_to_flush()`, which is not trivial inside the `MAX()` macro.

robn replied:
Are you thinking to just make it more visible, by splitting the lines out? Or to actually not call it, and do some kind of more lightweight calculation in its place, or to decide whether to call it?

amotin replied:
Making it obvious to compiler that the function should not be called twice due to two mentions in the macro.

robn replied:
Oh yeah, that's a thing isn't it. Sigh. Thanks, fixed.

amotin commented on module/zfs/spa_log_spacemap.c:
```diff
@@ -847,14 +999,27 @@ spa_flush_metaslabs(spa_t *spa, dmu_tx_t *tx)
+				pool_condense_stat_t *pcns =
+				    &spa->spa_condense_stats
+				    [POOL_CONDENSE_LOG_SPACEMAP];
+				pcns->pcns_processed--;
```
Shouldn't `pcns_processed` grow?

robn replied:
Indeed. Why did I do that?

amotin reviewed (commented):
CI is unhappy:
```
05:52:16.23 invalid condense type 'log-spacemap'
05:52:16.23 usage:
05:52:16.23     condense [-t <target>] [-c | -w] [-a | <pool> ...]
05:52:16.23 ERROR: zpool condense -t log-spacemap -w logsm_flushall exited 2
```

amotin commented:
```
  [ 6457.250147] VERIFY0(nvlist_add_nvlist(nvl, name, val)) failed (22)
  [ 6457.251511] PANIC at fnvpair.c:206:fnvlist_add_nvlist()
  [ 6457.253276] Showing stack for process 1745664
  [ 6457.254497] CPU: 0 UID: 0 PID: 1745664 Comm: zpool Tainted: P           OE  X   ------  ---  6.12.0-124.55.3.el10_1.x86_64 #1 PREEMPT(voluntary) 
  [ 6457.254711] Tainted: [P]=PROPRIETARY_MODULE, [O]=OOT_MODULE, [E]=UNSIGNED_MODULE, [X]=AUX
  [ 6457.254725] Hardware name: QEMU Ubuntu 24.04 PC (Q35 + ICH9, 2009), BIOS 1.16.3-debian-1.16.3-2 04/01/2014
  [ 6457.255007] Call Trace:
  [ 6457.255347]  <TASK>
  [ 6457.255397]  dump_stack_lvl+0x4e/0x70
  [ 6457.257190]  spl_panic+0xf4/0x119 [spl]
  [ 6457.260244]  root_vdev_actions_getprogress+0x191/0x210 [zfs]
  [ 6457.307712]  vdev_config_generate+0x51e/0xa10 [zfs]
  [ 6457.308033]  spa_config_generate+0x26f/0x510 [zfs]
  [ 6457.308242]  spa_open_common+0x195/0x3e0 [zfs]
  [ 6457.308493]  ? __lruvec_stat_mod_folio+0xa7/0xd0
  [ 6457.309330]  spa_get_stats+0x4c/0x1e0 [zfs]
  [ 6457.309979]  zfs_ioc_pool_stats+0x3e/0x90 [zfs]
  [ 6457.310191]  zfsdev_ioctl_common+0x64a/0x6e0 [zfs]
  [ 6457.310368]  zfsdev_ioctl+0x53/0xe0 [zfs]
  [ 6457.310530]  __x64_sys_ioctl+0x94/0xd0
```

amotin commented on module/zfs/spa_log_spacemap.c:
```diff
@@ -747,10 +756,116 @@ spa_log_exceeds_memlimit(spa_t *spa)
+		    &spa->spa_condense_stats[SPA_CONDENSE_LOG_SPACEMAP];
+		memset(scns, 0, sizeof (spa_condense_stat_t));
+		scns->scns_start_time = gethrestime_sec();
+		scns->scns_total = spa_log_sm_unflushed_metaslabs(spa);
```
Would `txg` ever be not the current, this would not be correct.

robn replied:
It's true. I've added a comment and assertion to this effect.

behlendorf commented on man/man8/zpool-condense.8:
```diff
@@ -0,0 +1,68 @@
+.Sy condense .
+This will return background updates to their normal rate.
+.It Fl w , -wait
+Wait until the condense has completed before returning.
```
We've had requests for more examples in the man pages, let's include one.  How about something like this (modeled after `zpool-scrub(8)`).

```
# zpool condense -t log_spacemap
# zpool status
...
<example condense output>
...
```

robn replied:
Right on, done.

behlendorf reviewed (commented):
Functionally this looks great.  But before merging this can you add a few tests to address these gaps in the test coverage.

- Basic tests added under `tests/functional/cli_root/zpool_condense/`.  Minimally we want to exercise each command line option (--all, --cancel, --wait, --type), plus a negative test which verified invalid arguments are rejected.
- Add a `zpool wait -t condense` test to `/tests/functional/cli_root/zpool_wait/`.

robn committed:
zts: zpool-condense sanity tests

Sponsored-by: TrueNAS
Signed-off-by: Rob Norris <rob.norris@truenas.com>

robn commented:
Last push addresses the two inline comments, should finish the test tomorrow.

robn commented:
Alright, last push adds some basic `zpool condense` tests.

I looked at how scrub tests control what's happening from the tests, through a "pause" tuneable, but it didn't really seem to fit well because condense is more of an acceleration signal + stats package, vs an actual function.

So instead I added another condense type, `debug`, that is only compiled when `ZFS_DEBUG` is set. All it does is issue a series of delay tasks to bump the stats every second, and "complete" the job after 10s. The fixed time makes it easier to work to test the actual machinery than to do it via an actual condense type, which by design requires a lot of work to do and tries to do it all fast.

Obviously, adding a whole new thing to support just testing is always worth raising an eyebrow at, but I've tried to keep it very well contained and commented. And it did pay off, it helped me see that I'd completely forgotten to add locks to the stats, so that's all set up too.

robn commented:
Assuming the tests pass, I think I'm done here. But I should flag a naming/design thing that has been on my mind.

I don't think "condense" is a very pleasant term for whatever this is. At least, we do have some other condense things (indirect, metaslab, livelist) that aren't quite related, and it being a somewhat unusual word might make it easy to imagine they're related.

The actual thing its aimed at is things that have a journal or log that gets pulled into the "main" data from time to time. When we started on it, the obvious next target was expected to be the dedup log, but after that, maybe nothing?

Now ultimately all this is a facility for progress counting. It's really just a generalisation of other progress facilities: scrub, rebuild, initialize, trim. Not the same of course, but its not a huge leap to imagine they could be folded back into a single structure. I'm not proposing that, but there's a future where that happens.

Further, that future might actually be one where make the idea of a long-running "job" into a first-class concept. By this I mean something multi-step, multi-txg, that maybe lives across export/import cycles - exactly what scrub is right now. There's certainly plenty of background "work" that happens in ZFS, and in large part they all manage themselves. I think there's some kind of case for a central scheduling/prioirtising engine that is keeping track of where everything is at, and perhaps giving the operator tools to manipulate it, reprioritise jobs, etc.

I'm not arguing for that _now_ - its a big job, lots of research and experimentation wanted. But it is jumbled into my thinking, because is introducing the start of a progress-reporting facility called "condense" really a good idea if it just adds another weird term and a whole separate mechanism into the mix?

Maybe I'd care less if it didn't just say "condense" all over the tooling and API, and so need to be supported into the future. Maybe it doesn't matter, maybe the all-singing all-dancing version just keeps the old names as aliases or shims over the new thing. The new thing which didn't exist three paragraphs ago!

I have no question. And maybe there's not even a concern. I just can't be sure I'm not tying our hands in some annoying way, and that bugs me a bit!

behlendorf commented:
> But I should flag a naming/design thing that has been on my mind.

I've been giving this some thought and I get where your coming from.  It's an awful lot of machinery to add solely to provide an interface to flush the spacemap logs.  Sure, we may we want to extend it to the dedup log, what's after that I'm not sure.  I suppose if we wanted a more inclusive name, and to avoid adding another subcommand, we could extend `zpool sync` to cover condensing the spacemap logs.  But I'm not sure I love that either.  Let me mull it over a bit more, at the moment I'm leaning towards sticking with "condense".  It's descriptive, and I wouldn't be at all surprised if we end up finding data structures we want to add to it.

behlendorf reviewed (approved):
Well after sitting with this for a bit I'm on board with "condense" for the naming.  Rob if you can look in to the `cli_root/zpool_wait/zpool_wait_initialize_basic` failure all the builders hit I think this is ready.  

```
cli_root/zpool_wait/zpool_wait_initialize_basic (57 KiB)
  [2026-06-04T05:48:41.204227] Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_wait/zpool_wait_initialize_basic (run as root) [00:00] [FAIL]
  05:48:40.59 SUCCESS: set_tunable64 INITIALIZE_CHUNK_SIZE 2048
  05:48:40.60 SUCCESS: mkdir /var/tmp/testdir
  05:48:40.98 SUCCESS: mkfile 256M /var/tmp/testdir/file_vdev
  05:48:40.99 cannot create 'testpool': pool already exists
  05:48:40.99 ERROR: zpool create -f testpool /var/tmp/testdir/file_vdev exited 1
```

gmelikov commented:
FWIW, +1 for condense :)

robn commented:
Last push should fix it.

The "debug" condense target has a quirk, that I described in the comment above `spa_condense_debug_task`. Basically, a `spa_condense_debug_task` task may exist on the delay taskq when the condense completes or is cancelled, until the next tick in 1s, when it is called, sees the condense is finished, and exits without re-arming itself.

I've done it this way because otherwise I would have to explicitly cancel it, which would mean adding an extra `taskqid_t` field to `spa_t` just for this debug function.

While that debug task is outstanding, there's a hold on the pool, which prevents it being destroyed. That's what `zpool_wait_initialize_basic` was hitting: the previous test, `zpool_wait_condense`, had not destroyed the pool, and it tried to `create -f` before the delay task has fired and released its hold.

I'd tried to work around it by putting `log_must_busy zpool destroy $TESTPOOL` in the `cleanup` function, but it seems like my `log_onexit` function is only call if the test fails? Which seems wrong, at least, its not how I've thought it worked for a good long while, but that's what I'm seeing. (I didn't do the context switch to read `logapi` to confirm it right now).

So last push sticks `log_must_busy zpool destroy $TESTPOOL` at the bottom of the test proper, with a comment. That seems to sort it out.

At this point, probably I should it the right way and add the `spa_t` field and clean up properly. I was still surprised by the tests though, so idk.

behlendorf commented:
@robn thanks for running the test failure down.  Yeah, it's a debug quirk I think we can live with in the interests of getting this across the finish line.  If you want to circle back and the `spa_t` field for this I'm all for it, but it doesn't need to block this.

behlendorf commented:
Almost there.  @robn it looks like this change introduced a `zloop` hang which needs to be sorted out.  The [first CI run](https://github.com/openzfs/zfs/actions/runs/27178711749/attempts/1?pr=16747) hit the hard time limit, I resubmitted the job and it looks like it's hung as well.  So the good news is it's reproducible!

robn commented:
Guess I'm doing it the right way right now then. Thanks for the patience! On it right now, hopefully will be back shortly with a patch.

robn commented:
Last push(es) rework the debug target to track the delay task properly and cancel it, as discussed. Seems better!

It occurred to me as I was doing that ztest doesn't use that target; it does the actual spacemap condense. So that's not gonna be it directly, but I do wonder if its a similar shape - condense task running at whatever shutdown it does? Or, semirelatedly, some resource held open. Never actually used zloop, so awesome - another side quest! :sweat_smile:

robn commented:
zloop with the same args as used in CI consistently runs locally to happy completion in ~12mins, so I'll have to try something else. I notice the CI one itself runs on top of a pool with autotrim on, so maybe that's actually where the hang is coming from? No idea. I'll try to mimic that environment more closely tomorrow, but I'll happily take guidance too.

behlendorf commented:
Well normally I'd suggest downloading the artifacts from the CI to debug this.  But in this case since it was cancelled and didn't technically fail it looks like they didn't get uploaded.  If you update the `if: failure()` lines in the `.github/workflows/zfs-qemu.yml` workflow to `if: failure() || cancelled()` I think you'll be able to grab the logs.  Hopefully that will shed some light on what's happening.

robn commented:
@behlendorf thanks, good tip. I've pushed that now to see how it goes; much better then faffing about locally (which I haven't gotten around to yet, as you see). Fingers crossed!

robn commented:
Trying again. zloop only saves the logs from its current run if there's a core, so those don't included. If I've done it right, this run should grab the in-flight ones too.

I've also thrown in the kernel log, since the ZFS build is ASAN/UBSAN/kmem enabled, For all I know its sitting there in a panic state of some sort.

I did manage to do a real test locally on top of a real pool, and it was fine. I'll see what debug trimmings I can enable.

robn commented:
There it is, in the "main" `ztest.out`, not the core ones that we copy aside:

```
=================================================================
==72724==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x7f7698afd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f7699167b74 in umem_alloc lib/libspl/include/umem.h:95
    #2 0x7f7699169f51 in nv_mem_zalloc module/nvpair/nvpair.c:208
    #3 0x7f7699170789 in nvlist_xalloc module/nvpair/nvpair.c:607
    #4 0x7f769917fdcc in fnvlist_alloc module/nvpair/fnvpair.c:55
    #5 0x7f7696f4f24a in root_vdev_actions_getprogress module/zfs/vdev_label.c:450
    #6 0x7f7696f57cf9 in vdev_config_generate module/zfs/vdev_label.c:661
    #7 0x7f7696e648ed in spa_config_generate module/zfs/spa_config.c:431
    #8 0x7f7696e462ff in spa_open_common module/zfs/spa.c:6566
    #9 0x7f7696e4792a in spa_get_stats module/zfs/spa.c:6831
    #10 0x7f76969752bb in show_pool_stats lib/libzpool/util.c:141
    #11 0x56504718ea93 in dump_zpool cmd/zdb/zdb.c:9216
    #12 0x56504711c54e in main cmd/zdb/zdb.c:10626
    #13 0x7f769482a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #14 0x7f769482a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #15 0x56504711dcc4 in _start (/usr/sbin/zdb+0xb5cc4) (BuildId: c2e52351d6ef0c00dcacd9e4f1c8eb110e401208)

Indirect leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f7698afd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f7699167b74 in umem_alloc lib/libspl/include/umem.h:95
    #2 0x7f769916c574 in nv_priv_alloc module/nvpair/nvpair.c:240
    #3 0x7f7699170770 in nvlist_xalloc module/nvpair/nvpair.c:604
    #4 0x7f769917fdcc in fnvlist_alloc module/nvpair/fnvpair.c:55
    #5 0x7f7696f4f24a in root_vdev_actions_getprogress module/zfs/vdev_label.c:450
    #6 0x7f7696f57cf9 in vdev_config_generate module/zfs/vdev_label.c:661
    #7 0x7f7696e648ed in spa_config_generate module/zfs/spa_config.c:431
    #8 0x7f7696e462ff in spa_open_common module/zfs/spa.c:6566
    #9 0x7f7696e4792a in spa_get_stats module/zfs/spa.c:6831
    #10 0x7f76969752bb in show_pool_stats lib/libzpool/util.c:141
    #11 0x56504718ea93 in dump_zpool cmd/zdb/zdb.c:9216
    #12 0x56504711c54e in main cmd/zdb/zdb.c:10626
    #13 0x7f769482a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #14 0x7f769482a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #15 0x56504711dcc4 in _start (/usr/sbin/zdb+0xb5cc4) (BuildId: c2e52351d6ef0c00dcacd9e4f1c8eb110e401208)

SUMMARY: AddressSanitizer: 80 byte(s) leaked in 2 allocation(s).
Registers:
  RAX: 0x0000000000000000  RDX: 0x00000fef52558200  RCX: 0x0000000000000000
  RBX: 0x0000000000000002  RSI: 0x00007f7692b00000  RDI: 0x00007f7692b00050
  RBP: 0x00007ffc4fe0d350  RSP: 0x00007ffc4fe0c8d0   R8: 0x0000000000000000
   R9: 0x00000feed2560000  R10: 0x0000000000000000  R11: 0x0000000000000000
  R12: 0x00007f7692b00050  R13: 0x00007f7692750c20  R14: 0x0000000000000006
  R15: 0x00007f7692b009e0  RIP: 0x00007f76975422be
Call trace:
  [0x00007f76975422be] libspl_backtrace+0x12e
  [0x00005650471498b7] sig_handler+0xa7
  [0x00007f7694845330] __sigaction+0x50
  [0x00007f769489eb2c] pthread_kill+0x11c
  [0x00007f769484527e] gsignal+0x1e
  [0x00007f76948288ff] abort+0xdf
  [0x00007f7698b1fc30] _ZN11__sanitizer5AbortEv+0x30
  [0x00007f7698b307cd] _ZN11__sanitizer3DieEv+0x4d
  [0x00007f7698b3c9cf] _ZN6__lsan11HandleLeaksEv+0x1f
  [0x00007f7698b3a105] _ZN6__lsan11DoLeakCheckEv+0x125
  [0x00007f7694847382] __cxa_finalize+0xc2
  [0x00007f7698a3b7b7] __do_global_dtors_aux+0x27
  [0x00007f76992690f2] ???
  [0x00007f769926d578] __nptl_change_stack_perm+0x1288
  [0x00007f7694847a76] erand48_r+0x1d6
  [0x00007f7694847bbe] exit+0x1e
  [0x00007f769482a1d1] __libc_init_first+0x91
  [0x00007f769482a28b] __libc_start_main+0x8b
  [0x000056504711dcc5] _start+0x25

ASSERT at lib/libspl/mutex.c:58:mutex_enter()
VERIFY0(pthread_mutex_lock(&mp->m_lock)) failed (22)
  PID: 72724     COMM: zdb
  TID: 72724     NAME: zdb
Registers:
  RAX: 0x0000000000000000  RDX: 0x00000fef52558400  RCX: 0x0000000000000000
  RBX: 0x0000000000000002  RSI: 0x00007f7692b01000  RDI: 0x00007f7692b01050
  RBP: 0x00007ffc4fe0d0c0  RSP: 0x00007ffc4fe0c640   R8: 0x0000000000000000
   R9: 0x00000feed2560200  R10: 0x0000000000000000  R11: 0x0000000000000000
  R12: 0x00007f7692b01050  R13: 0x00000feed24ea180  R14: 0x00007f7694a044e0
  R15: 0x00007f7692b019e0  RIP: 0x00007f76975422be
Call trace:
  [0x00007f76975422be] libspl_backtrace+0x12e
  [0x00007f7697541ee3] libspl_assertf+0x263
  [0x00007f769754a848] mutex_enter+0xa8
  [0x00007f7696977dd6] zfs_dbgmsg_print+0x26
  [0x00005650471498bc] sig_handler+0xac
  [0x00007f7694845330] __sigaction+0x50
  [0x00007f769489eb2c] pthread_kill+0x11c
  [0x00007f769484527e] gsignal+0x1e
  [0x00007f76948288ff] abort+0xdf
  [0x00007f7698b1fc30] _ZN11__sanitizer5AbortEv+0x30
  [0x00007f7698b307cd] _ZN11__sanitizer3DieEv+0x4d
  [0x00007f7698b3c9cf] _ZN6__lsan11HandleLeaksEv+0x1f
  [0x00007f7698b3a105] _ZN6__lsan11DoLeakCheckEv+0x125
  [0x00007f7694847382] __cxa_finalize+0xc2
  [0x00007f7698a3b7b7] __do_global_dtors_aux+0x27
  [0x00007f76992690f2] ???
  [0x00007f769926d578] __nptl_change_stack_perm+0x1288
  [0x00007f7694847a76] erand48_r+0x1d6
  [0x00007f7694847bbe] exit+0x1e
  [0x00007f769482a1d1] __libc_init_first+0x91
  [0x00007f769482a28b] __libc_start_main+0x8b
  [0x000056504711dcc5] _start+0x25
Registers:
  RAX: 0x0000000000000000  RDX: 0x00000fef52558600  RCX: 0x0000000000000000
  RBX: 0x0000000000000002  RSI: 0x00007f7692b02000  RDI: 0x00007f7692b02050
  RBP: 0x00007ffc4fe0c890  RSP: 0x00007ffc4fe0be10   R8: 0x0000000000000000
   R9: 0x00000feed2560400  R10: 0x0000000000000000  R11: 0x0000000000000000
  R12: 0x00007f7692b02050  R13: 0x00007f7692750e20  R14: 0x0000000000000006
  R15: 0x00007f7692b029e0  RIP: 0x00007f76975422be
Call trace:
  [0x00007f76975422be] libspl_backtrace+0x12e
  [0x00005650471498b7] sig_handler+0xa7
  [0x00007f7694845330] __sigaction+0x50
  [0x00007f769489eb2c] pthread_kill+0x11c
  [0x00007f769484527e] gsignal+0x1e
  [0x00007f76948288ff] abort+0xdf
  [0x00007f7696968098] libspl_assertf.cold+0x0a
  [0x00007f769754a848] mutex_enter+0xa8
  [0x00007f7696977dd6] zfs_dbgmsg_print+0x26
  [0x00005650471498bc] sig_handler+0xac
  [0x00007f7694845330] __sigaction+0x50
  [0x00007f769489eb2c] pthread_kill+0x11c
  [0x00007f769484527e] gsignal+0x1e
  [0x00007f76948288ff] abort+0xdf
  [0x00007f7698b1fc30] _ZN11__sanitizer5AbortEv+0x30
  [0x00007f7698b307cd] _ZN11__sanitizer3DieEv+0x4d
  [0x00007f7698b3c9cf] _ZN6__lsan11HandleLeaksEv+0x1f
  [0x00007f7698b3a105] _ZN6__lsan11DoLeakCheckEv+0x125
  [0x00007f7694847382] __cxa_finalize+0xc2
  [0x00007f7698a3b7b7] __do_global_dtors_aux+0x27
  [0x00007f76992690f2] ???
  [0x00007f769926d578] __nptl_change_stack_perm+0x1288
  [0x00007f7694847a76] erand48_r+0x1d6
  [0x00007f7694847bbe] exit+0x1e
  [0x00007f769482a1d1] __libc_init_first+0x91
  [0x00007f769482a28b] __libc_start_main+0x8b
  [0x000056504711dcc5] _start+0x25

```

So there's a leak somewhere, but also this is `zdb` crashing, not `ztest`. So no core, so `zloop` doesn't notice it, so it all hangs up.

I'll get back to this on Monday at the latest. I'll fix the bug, but I'll also lift out and polish the CI side a bit. Something like what I did already, but I think I wanna adjust `zloop.sh` itself so it doesn't wait longer than the timeout, and it looks harder for the output.

robn commented:
Alright, last push should sort that out. Simple bug, a mistake I make often: `fnvlist_add_nvlist(nv, name, cnv)` does not take ownership of `cnv`, so you have to `fnvlist_free()`. No big deal, just seemed bigger because it took a while to get that info out.

If the tests pass, I'll push again dropping the CI hacks, and then that'll be good to go. I've put the CI hacks on a list; I'll look to improve that in a future PR.

robn commented:
That ran clean, two unrelated failures. Last push dropped the zloop/CI commit, so this should be the final product!

amotin commented on include/sys/spa_impl.h:
```diff
@@ -217,6 +219,23 @@ typedef enum spa_config_source {
+	SPA_CONDENSE_LOG_SPACEMAP = 0,
+#ifdef ZFS_DEBUG
+	SPA_CONDENSE_DEBUG,
+#endif
```
This will make `SPA_CONDENSE_TYPES` change depending on `ZFS_DEBUG`.  Though might be not a big deal if the numbers won't leak to user-space.

robn replied:
Yeah, this is just the index for `spa_condense_stats`. Its deliberately not exposed to userspace, the types are in the API are all stringy names.

amotin commented on cmd/zpool/zpool_main.c:
```diff
@@ -8850,6 +8870,151 @@ zpool_do_trim(int argc, char **argv)
+			continue;
+#endif
+		err = zpool_condense(zhp, cb->cmd, condense_types[i]);
+		if (err != 0)
```
I suppose for pools without log spacemaps enabled we'll get `ENOSUP` here, even though it would effectively be a NOP.

robn replied:
Yeah, good though. Made it treat `ENOTSUP` here as "success", so we don't skip any remainng types.

amotin commented on include/sys/spa_impl.h:
```diff
@@ -217,6 +219,23 @@ typedef enum spa_config_source {
+	uint64_t scns_end_time;		/* time_t */
+	uint64_t scns_processed;	/* items processed */
+	uint64_t scns_total;		/* total items to process */
+	kmutex_t scns_lock;		/* protects all of the above */
```
`scns_lock` seems to be unused. It seems `spa_condense_stats_lock` protects the whole array.

robn replied:
Ahh good catch. Yes, I was of two minds about a single big lock vs a per-type lock, because I didn't want two things running at the same time and updating stats to slow each other down. I decided we don't have two things, so a single lock would be fine for now, but I forgot to remove `scns_lock`. Done now, thanks!

behlendorf reviewed (approved):
Ah, okay.  Thanks for running down that loose end.

behlendorf commented:
Merged.  Thanks @robn for iterating on this, I'm happy to see it get across the finish line.  I'm sure you are too!

robn commented:
You've no idea. Thanks all for the patience and hard work!
